subsys/usb/class/audio/audio.c

```diff
@@ -10,3 +10,605 @@
  *
  * Driver for USB Audio device class driver
  */
+
+#include <kernel.h>
+#include <usb/usb_common.h>
+#include <usb/usb_device.h>
+#include <usb_descriptor.h>
+#include <usb/usbstruct.h>
+#include <usb/class/usb_audio.h>
+
+#include <sys/byteorder.h>
+#include <net/buf.h>
+
+/* TODO: Kconfig options for below defines */
+#define AUDIO_EP_SIZE	192
+
+#define LOG_LEVEL CONFIG_USB_AUDIO_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(usb_audio);
+
+static void audio_buffer_destroyed(struct net_buf *buf);
+
+NET_BUF_POOL_FIXED_DEFINE(audio_data_pool, 5,
+			  AUDIO_EP_SIZE,
+			  audio_buffer_destroyed);
+
+/* Device data structure */
+struct usb_audio_dev_data_t {
+	/* Number of configured channels */
+	u8_t channels;
+
+	/* Supported features mask */
+	u16_t features;
+
+	const struct audio_ops *ops;
+
+	struct controls *controls;
+
+	struct usb_dev_data common;
+
+	bool rx_tx_enable;
+};
+
+static sys_slist_t usb_audio_data_devlist;
+
+/**
+ * @brief Fill the USB Audio descriptor
+ *
+ * This macro fills USB descriptor for specific type of device
+ * (Heahphones or Microphone) depending on given parameters.
+ *
+ * @param [in] dev	Device type. Must be HEADPHONES/MICROPHONE
+ * @param [in] i0	First interface ID
+ * @param [in] e0	First entity ID
+ * @param [in] it_type	Input terminal type
+ * @param [in] ot_type	Output terminal type
+ * @param [in] ep_addr	Endpoint address associated with audio function
+ *
+ */
+#define DEFINE_AUDIO_DESCRIPTOR(dev, link, it_type, ot_type, ep_addr)			\
```

Comment:
That is not pretty, you could split it to few smaller chunks.

Reply:
Will do something like: 
```
#define DEFINE_AUDIO_HP_DESC(dev, link)
        DEFINE_AUDIO_DESCRIPTOR(dev, USB_STREAMING, OUT_HEADPHONES, <addr_from_dt>) 

...

UTIL_LISTIFY(CONFIG_NUM_OF_HP_DEVICES, DEFINE_AUDIO_HP_DESC, _)  //something like this
```
<addr_from_dt> -> see comment in Kconfig